packages/muelu/src/Utils/MueLu_UtilitiesBase_decl.hpp

```diff
@@ -300,52 +300,136 @@ namespace MueLu {
       if(bA == Teuchos::null) {
         RCP<const Map> rowMap = rcpA->getRowMap();
         diag = Xpetra::VectorFactory<Scalar,LocalOrdinal,GlobalOrdinal,Node>::Build(rowMap,true);
-        ArrayRCP<Scalar> diagVals = diag->getDataNonConst(0);
-        Teuchos::Array<Scalar> regSum(diag->getLocalLength());
-        Teuchos::ArrayView<const LocalOrdinal> cols;
-        Teuchos::ArrayView<const Scalar> vals;
 
-        std::vector<int> nnzPerRow(rowMap->getLocalNumElements());
-
-        //FIXME 2021-10-22 JHU   If this is called with doReciprocal=false, what should the correct behavior be?  Currently,
-        //FIXME 2021-10-22 JHU   the diagonal entry is set to be the sum of the absolute values of the row entries.
-
-        const Magnitude zeroMagn = TST::magnitude(zero);
-        Magnitude avgAbsDiagVal = TST::magnitude(zero);
-        int numDiagsEqualToOne = 0;
-        for (size_t i = 0; i < rowMap->getLocalNumElements(); ++i) {
-          nnzPerRow[i] = 0;
-          rcpA->getLocalRowView(i, cols, vals);
-          diagVals[i] = zero;
-          for (LocalOrdinal j = 0; j < cols.size(); ++j) {
-            regSum[i] += vals[j];
-            const Magnitude rowEntryMagn = TST::magnitude(vals[j]);
-            if (rowEntryMagn > zeroMagn)
-              nnzPerRow[i]++;
-            diagVals[i] += rowEntryMagn;
-            if (static_cast<size_t>(cols[j]) == i)
-              avgAbsDiagVal += rowEntryMagn;
-          }
-          if (nnzPerRow[i] == 1 && TST::magnitude(diagVals[i])==1.)
-            numDiagsEqualToOne++;
-        }
-        if (useAverageAbsDiagVal)
-          tol = TST::magnitude(100 * Teuchos::ScalarTraits<Scalar>::eps()) * (avgAbsDiagVal-numDiagsEqualToOne) / (rowMap->getLocalNumElements()-numDiagsEqualToOne);
-        if (doReciprocal) {
-          for (size_t i = 0; i < rowMap->getLocalNumElements(); ++i) {
-            if (replaceSingleEntryRowWithZero && nnzPerRow[i] <= static_cast<int>(1))
-              diagVals[i] = zero;
-            else if ((diagVals[i] != zero) && (TST::magnitude(diagVals[i]) < TST::magnitude(two*regSum[i])))
-              diagVals[i] = one / TST::magnitude((two*regSum[i]));
-            else {
-              if(TST::magnitude(diagVals[i]) > tol)
-                diagVals[i] = one / diagVals[i];
-              else {
-                diagVals[i] = valReplacement;
-              }
-            }
-          }
-        }
+	if(rowMap->lib() == Xpetra::UnderlyingLib::UseTpetra) {
```

Comment:
Why is the distinction between Tpetra and Epetra needed? Epetra does support the local matrix stuff, no?

Reply:
One can get a localDeviceView?

Reply:
https://github.com/trilinos/Trilinos/blob/78cf276df2e4a91296100d0ee25c31d93d18e976/packages/xpetra/src/CrsMatrix/Xpetra_EpetraCrsMatrix.hpp#L1230-L1237

Reply:
I honestly did not want to have to check and do no think that there is a real benefit to not run the old code when Epetra is being used...